src/internal/packager/images/pull.go

```diff
@@ -27,329 +32,329 @@ import (
 
-	spinner := message.NewProgressSpinner("Loading metadata for %d images. %s", imageCount, longer)
+	if err := helpers.CreateDirectory(cfg.DestinationDirectory, helpers.ReadExecuteAllWriteUser); err != nil {
+		return nil, fmt.Errorf("failed to create image path %s: %w", cfg.DestinationDirectory, err)
+	}
+
+	cranePath, err := clayout.FromPath(cfg.DestinationDirectory)
+	if err != nil {
+		cranePath, err = clayout.Write(cfg.DestinationDirectory, empty.Index)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	spinner := message.NewProgressSpinner("Fetching info for %d images. %s", imageCount, longer)
 	defer spinner.Stop()
 
 	logs.Warn.SetOutput(&message.DebugWriter{})
 	logs.Progress.SetOutput(&message.DebugWriter{})
 
-	metadataImageConcurrency := helpers.NewConcurrencyTools[ImgInfo, error](len(i.ImageList))
+	eg, _ := errgroup.WithContext(ctx)
+	eg.SetLimit(10)
 
-	defer metadataImageConcurrency.Cancel()
+	var fetchedLock, shaLock sync.Mutex
+	shas := make(map[string]bool)
+	opts := append(CommonOpts(cfg.Arch), crane.WithContext(ctx))
 
-	spinner.Updatef("Fetching image metadata (0 of %d)", len(i.ImageList))
+	var fetched = map[transform.Image]v1.Image{}
 
-	// Spawn a goroutine for each image to load its metadata
-	for _, refInfo := range i.ImageList {
-		// Create a closure so that we can pass the src into the goroutine
-		refInfo := refInfo
-		go func() {
+	var counter, totalBytes atomic.Int64
 
-			if metadataImageConcurrency.IsDone() {
-				return
-			}
+	for _, refInfo := range cfg.References {
+		refInfo := refInfo
+		eg.Go(func() error {
+			idx := counter.Add(1)
+			spinner.Updatef("Fetching image info (%d of %d)", idx, imageCount)
 
-			actualSrc := refInfo.Reference
-			for k, v := range i.RegistryOverrides {
+			ref := refInfo.Reference
+			for k, v := range cfg.RegistryOverrides {
 				if strings.HasPrefix(refInfo.Reference, k) {
-					actualSrc = strings.Replace(refInfo.Reference, k, v, 1)
+					ref = strings.Replace(refInfo.Reference, k, v, 1)
 				}
 			}
 
-			if metadataImageConcurrency.IsDone() {
-				return
-			}
-
-			img, hasImageLayers, err := i.PullImage(actualSrc, spinner)
-			if err != nil {
-				metadataImageConcurrency.ErrorChan <- fmt.Errorf("failed to pull %s: %w", actualSrc, err)
-				return
-			}
+			var img v1.Image
+			var desc *remote.Descriptor
 
-			if metadataImageConcurrency.IsDone() {
-				return
+			// load from local fs if it's a tarball
+			if strings.HasSuffix(ref, ".tar") || strings.HasSuffix(ref, ".tar.gz") || strings.HasSuffix(ref, ".tgz") {
+				img, err = crane.Load(ref, opts...)
+				if err != nil {
+					return fmt.Errorf("unable to load %s: %w", refInfo.Reference, err)
+				}
+			} else {
+				reference, err := name.ParseReference(ref)
+				if err != nil {
+					return fmt.Errorf("failed to parse reference: %w", err)
+				}
+				desc, err = crane.Get(ref, opts...)
+				if err != nil {
+					if strings.Contains(err.Error(), "unexpected status code 429 Too Many Requests") {
+						cancel()
```

Comment:
Why is cancel being called here? Any error returned from an error group will cancel the context used to create it.

Reply:
I should have tested it, but I was under the impression context cancellation was a one way data flow, where parent cancellations would kill dependent contexts, but not the other way around.